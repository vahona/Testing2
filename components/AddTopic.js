import React, { useState } from "react";
export default function AddTopic({ addNewTopic }) {
  const [topicTitle, setTopicTitle] = useState("");
  return (
    <section>
      <header>
        <h4 data-testid="add-topic-title">Add a topic</h4>
      </header>
      <article>
        <form
          onSubmit={(e) => {
            e.preventDefault();
            e.target.reset();
            addNewTopic(topicTitle);
          }}
          className="add-topic"
        >
          <input
            data-testid="topic-title-input"
            required
            type="text"
            name="topic"
            placeholder="Type your topic idea here"
            value={topicTitle}
            onChange={(e) => setTopicTitle(e.target.value)}
          />
          <button type="submit" data-testid="submit-button">
            Submit
          </button>
        </form>
      </article>
    </section>
  );
}
