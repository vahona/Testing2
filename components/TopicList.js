import React from "react";
export default function TopicList({
  title,
  topics,
  archiveTopic,
  deleteTopic,
  upvoteTopic,
  downvoteTopic,
}) {
  if (!topics) {
    console.error("No topics passed");
    return null;
  }
  return (
    <section className="next-topic">
      <header>
        <h4 data-testid="topic-list-title">{title}</h4>
      </header>
      <section className="next-topic-container">
        {!topics.length && "Loading..."}
        {topics.map((topic) => {
          const discussedOnDate = new Date(Number(topic.discussedOn));
          return (
            <article key={topic.id}>
              {!topic.discussedOn ? (
                <button
                  onClick={() => archiveTopic(topic.id)}
                  className="archive"
                  data-testid="archive-button"
                >
                  <svg
                    className="w-6 h-6"
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
                    ></path>
                  </svg>
                </button>
              ) : (
                <button
                  onClick={() => deleteTopic(topic.id)}
                  className="delete"
                  data-testid="delete-button"
                >
                  <svg
                    className="w-6 h-6"
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
                    ></path>
                  </svg>
                </button>
              )}
              <h5 className="topic-text">{topic.title}</h5>
              {!topic.discussedOn ? (
                <div className="votes">
                  <button
                    onClick={() => upvoteTopic(topic.id)}
                    className="upvote"
                    data-testid="upvote-topic"
                  >
                    <svg
                      class="w-6 h-6"
                      fill="none"
                      stroke="currentColor"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        d="M14 10h4.764a2 2 0 011.789 2.894l-3.5 7A2 2 0 0115.263 21h-4.017c-.163 0-.326-.02-.485-.06L7 20m7-10V5a2 2 0 00-2-2h-.095c-.5 0-.905.405-.905.905 0 .714-.211 1.412-.608 2.006L7 11v9m7-10h-2M7 20H5a2 2 0 01-2-2v-6a2 2 0 012-2h2.5"
                      ></path>
                    </svg>
                  </button>
                  <span className="upvote-number">{topic.upvotes}</span>
                  <button
                    onClick={() => downvoteTopic(topic.id)}
                    className="downvote"
                    data-testid="down-topic"
                  >
                    <svg
                      class="w-6 h-6"
                      fill="none"
                      stroke="currentColor"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        d="M10 14H5.236a2 2 0 01-1.789-2.894l3.5-7A2 2 0 018.736 3h4.018a2 2 0 01.485.06l3.76.94m-7 10v5a2 2 0 002 2h.096c.5 0 .905-.405.905-.904 0-.715.211-1.413.608-2.008L17 13V4m-7 10h2m5-10h2a2 2 0 012 2v6a2 2 0 01-2 2h-2.5"
                      ></path>
                    </svg>
                  </button>
                  <span className="downvote-number">{topic.downvotes}</span>
                </div>
              ) : (
                <p>Discussed on {discussedOnDate.toLocaleDateString()}</p>
              )}
            </article>
          );
        })}
      </section>
    </section>
  );
}
