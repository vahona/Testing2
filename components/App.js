import React, { useState, useEffect } from 'react';
import AddTopic from './AddTopic';
import TopicList from './TopicList';

const url = `https://gist.githubusercontent.com/Pinois/93afbc4a061352a0c70331ca4a16bb99/raw/6da767327041de13693181c2cb09459b0a3657a1/topics.json`;

export default function App() {
	const [topics, setTopics] = useState([]);

	async function fetchTopics() {
    const res = await fetch(url);
    try {
      const data = await res.json();
      setTopics(data);
    } catch (e) {
      // console.log(e);
    }
  }

	useEffect(() => {
		fetchTopics();
	}, []);

	function addNewTopic(e) {
		e.preventDefault();
		const title = e.target.topic.value;
		if (!title) return;
		const newTopic = {
			title,
			upvotes: 0,
			downvotes: 0,
			id: Date.now(),
		};
		const newTopicList = [...topics, newTopic];
		setTopics(newTopicList);
		e.target.reset();
	}

	function archiveTopic(topicId) {
		const newTopicList = topics.map(topic => {
			if (topic.id === topicId) {
				return {
					...topic,
					discussedOn: Date.now(),
				};
			}
			return topic;
		});
		setTopics(newTopicList);
	}

	function upvoteTopic(topicId) {
		const newTopicList = topics.map(topic => {
			if (topic.id === topicId) {
				return {
					...topic,
					upvotes: topic.upvotes + 1,
				};
			}
			return topic;
		});
		setTopics(newTopicList);
	}

	function downvoteTopic(topicId) {
		const newTopicList = topics.map(topic => {
			if (topic.id === topicId) {
				return {
					...topic,
					downvotes: topic.downvotes + 1,
				};
			}
			return topic;
		});
		setTopics(newTopicList);
	}
	function deleteTopic(topicId) {
		const newList = topics.filter(topic => topic.id !== topicId);
		setTopics(newList);
	}

	function sortTopic(topicA, topicB) {
		const rateTopicA = topicA.upvotes - topicA.downvotes;
		const rateTopicB = topicB.upvotes - topicB.downvotes;
		return rateTopicB - rateTopicA;
	}

	return (
    <>
      <h1 data-testid="app-title"> Tea time topic ðŸ’¡</h1>
      <AddTopic addNewTopic={addNewTopic}></AddTopic>
      <TopicList
        archiveTopic={archiveTopic}
        deleteTopic={deleteTopic}
        upvoteTopic={upvoteTopic}
        downvoteTopic={downvoteTopic}
        title="Next Topics"
        topics={topics.filter((topic) => !topic.discussedOn).sort(sortTopic)}
      >
	  </TopicList>
      <TopicList
        archiveTopic={archiveTopic}
        deleteTopic={deleteTopic}
        upvoteTopic={upvoteTopic}
        downvoteTopic={downvoteTopic}
        title="Past Topics"
        topics={topics.filter((topic) => topic.discussedOn).sort(sortTopic)}
      ></TopicList>
    </>
  );
}
