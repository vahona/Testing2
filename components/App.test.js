import React from "react";
import App from "./App";
import { render, waitFor, fireEvent } from "@testing-library/react";
import fetchMock from "jest-fetch-mock";
import AddTopic from './AddTopic'
import TopicList from './TopicList'



fetchMock.enableMocks();
beforeEach(() => {
  fetchMock.resetMocks();
  //  console.error.mockClear();
  
});

const addNewTopic = jest.fn(() => {})

const downvoteTopic = jest.fn(() => {});

const archiveTopic = jest.fn(() => {});
const deleteTopic = jest.fn(() => {});
const upvoteTopic = jest.fn(() => {});

const sortTopic = jest.fn(() => {})


test("<App/> without data", () => {
  const { debug, queryByTestId, container } = render(<App />);
  expect(queryByTestId("app-title")).toBeTruthy();
  expect(container.firstChild).toMatchSnapshot();
  // debug();
});


const fakeTopics = [
  {
    id: "52e1bb8a-e848-4db0-8fb6-270d6ff139d2",
    upvotes: 3,
    title:
      "Some studies have shown that purpose is #1 for job satisfaction. What do you think?",
    downvotes: 25,
    discussedOn: "",
  },
  {
    id: "2c241a97-b4a5-4ea7-a29f-31714a48e36c",
    upvotes: 19,
    title:
      "People are quick to forget the good things done for them. Do you agree?",
    downvotes: 14,
    discussedOn: "1580417903018",
  },
];




test("<App/> with some fake fetch data", async () => {

	fetchMock.mockResponseOnce(JSON.stringify(fakeTopics));

	const { debug, queryByText, container } = render(<App />);
	await waitFor(() => queryByText(fakeTopics[0].title));
	expect(queryByText(fakeTopics[1].title)).toBeTruthy();
	expect(container.firstChild).toMatchSnapshot()
	debug();

});




test("<App addNewTopic={addNewTopic} /> add topic", () => {
    const { debug, queryByTestId, getByText } = render(<AddTopic addNewTopic={addNewTopic} />);
    expect(queryByTestId("submit-button")).toBeTruthy;

    fireEvent.click(getByText("Submit"));
    debug();
});




test("<App upvoteTopic={upvoteTopic}/> upvote", () => {
  const { debug, queryByTestId } = render(<TopicList topics={fakeTopics} upvoteTopic={upvoteTopic} />);
  expect(queryByTestId("upvote-topic")).toBeTruthy;

  fireEvent.click(queryByTestId("upvote-topic"));
    debug()
});


test("<App downvoteTopic={downvoteTopic}/> downvote", () => {
  const { debug, queryByTestId } = render(
    <TopicList topics={fakeTopics} downvoteTopic={downvoteTopic} />
  );
  expect(queryByTestId("downvote-topic")).toBeTruthy;

  fireEvent.click(queryByTestId("down-topic"));
  debug();
});




test("<App archiveTopic={archiveTopic}/> archive", () => {
  const { debug, queryByTestId } = render(
    <TopicList topics={fakeTopics} archiveTopic={archiveTopic} />
  );
  expect(queryByTestId("archive-button")).toBeTruthy;
  debug();
});



test("<App deleteTopic={deleteTopic}/> delete", () => {
  const { debug, queryByTestId } = render(
    <TopicList topics={fakeTopics} deleteTopic={deleteTopic} />
  );
  expect(queryByTestId("delete-button")).toBeTruthy;

   fireEvent.click(queryByTestId("delete-button"));
  debug();
});


test("<App  topics={topics.filter((topic) => topic.discussedOn).sort(sortTopic)}/>", () => {
  const { debug } = render(<TopicList topics={fakeTopics.filter((topic) => topic.discussedOn).sort(sortTopic)} />);

  const value = sortTopic(sortTopic);
  expect(sortTopic).toHaveBeenCalled()
  debug()
}
);


// install jest-fetch-mock
// require it
// create the mock data for app.js (tea-time-topics)
// create the test, and link the fetch mock response with the data
// use waitForElements to wait the data (mark your function as async)
// once you have that, test if the topics appears on the page
// start to do integrations tests to see if the functions work correctly


// Does the add topic work
// Does the upvote topic work
// Does the downvote topic work
// Does the archive topic work
// Does the delete topic work
// Does the sorting topic functionality works?
