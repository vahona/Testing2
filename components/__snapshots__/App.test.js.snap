// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> with some fake fetch data 1`] = `
<h1
  data-testid="app-title"
>
   Tea time topic ðŸ’¡
</h1>
`;

exports[`<App/> without data 1`] = `
<h1
  data-testid="app-title"
>
   Tea time topic ðŸ’¡
</h1>
`;
